Storage Blob swagger to typespec migration

catalinaperalta opened the pull request:
# Data Plane API Specification Update Pull Request

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.


<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## API Info: The Basics

Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories) 
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

mpodwysocki committed:
[Storage Blob] Updating routes

mpodwysocki committed:
[Storage Blob] Updating routes

mpodwysocki committed:
[Storage Blob] Updating routes

mpodwysocki committed:
[Storage Blob] Updating routes

mpodwysocki committed:
[Storage Blob] Updating routes

mpodwysocki committed:
[storage] Update models

mpodwysocki committed:
[storage] Update models

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] update the routes

mpodwysocki committed:
[storage blob] Fix dependencies

mpodwysocki committed:
[Storage Blob] Updating routes

mpodwysocki committed:
Update blob path parameter

cperaltah committed:
tsp format

cperaltah committed:
fix blob path param

cperaltah committed:
regen openapi

cperaltah committed:
add suppressions

cperaltah committed:
update code dependency

cperaltah committed:
remove package.json + lock

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
fix query params

cperaltah committed:
fix versioning

cperaltah committed:
refactor interface usage, fix operation ids

cperaltah committed:
tcgc cleanup

cperaltah committed:
import cleanup

cperaltah committed:
update namespace

cperaltah committed:
remove old versions

cperaltah committed:
update tspconfig.yaml

cperaltah committed:
fix versions, regen openapi

cperaltah committed:
add template

cperaltah committed:
clean up headers

cperaltah committed:
use template, regen

cperaltah committed:
remove operation id suppressions

cperaltah committed:
clientName clean up

cperaltah committed:
clean up headers and template

cperaltah committed:
add rust configs

cperaltah committed:
clean up and fix models

cperaltah committed:
fix routes

cperaltah committed:
regen openapi

cperaltah committed:
add content type to template

cperaltah committed:
revert - package.json update

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
rename endpoint param

cperaltah committed:
move version param to client init

cperaltah committed:
model corrections

cperaltah committed:
update routes and openapi

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
remove incorrect file changes

cperaltah committed:
undo package json changes

cperaltah committed:
add missing param

cperaltah committed:
fix tspconfig.yaml example dir

cperaltah committed:
specify page response

cperaltah committed:
add missing header

cperaltah committed:
more spec fixes

cperaltah committed:
more spec fixes - container interface

cperaltah committed:
specify client parameters

cperaltah committed:
regen openapi

cperaltah committed:
fix blob interface operations

cperaltah committed:
regen openapi

cperaltah committed:
fix pageblob operations

cperaltah committed:
move base blob interface route to each operation

cperaltah committed:
use template where possible

cperaltah committed:
fix appendblob operations

cperaltah committed:
fix NextMarker casing

cperaltah committed:
regen openapi

cperaltah committed:
update blockblob operations

cperaltah committed:
clean up more models and routes

cperaltah committed:
accuracy updates

cperaltah committed:
add placeholders for metadata and object replication headers

cperaltah committed:
update multipart/mixed content type

cperaltah committed:
update to 2025-01-05 api version

cperaltah committed:
reset swagger

cperaltah committed:
tspconfig.yaml updates

cperaltah committed:
example placeholder

cperaltah committed:
fix collection headers

cperaltah committed:
fixes

cperaltah committed:
revert wip

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
regen openapi

cperaltah committed:
add continuation token decorator

cperaltah committed:
simplify spec

cperaltah committed:
add linter suppression

cperaltah committed:
regen openapi

cperaltah committed:
clean up

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
regen openapi

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
use apiVersion decorator

cperaltah committed:
add typespec/xml to package.json

jhendrixMSFT committed:
remove duplicate x-ms-version headers

cperaltah committed:
spelling

cperaltah committed:
comment out @pagedResult decorator

cperaltah committed:
spelling + openapi regen

cperaltah committed:
update SignedIdentifiers

cperaltah committed:
clean up

cperaltah committed:
clean up

cperaltah committed:
remove containerName param from client

cperaltah committed:
add client customizations

cperaltah committed:
add other interfaces

jhendrixMSFT committed:
clean up some doc comments

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

jhendrixMSFT committed:
add missing x-ms-blob-type header

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
update lock file

cperaltah committed:
Merge branch 'feature/blob-tsp' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
regen openapi

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.Files.DataLake-Microsoft.BlobStorage](https://apiview.dev/Assemblies/Review/4f5ffebf4d1e4a0d9207fa4ec94a025f?revisionId=8dd85e2d89724c099e778e1988d64815)

cperaltah committed:
update blob type

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

jhendrixMSFT committed:
add missing content type for appendBlock

vincenttran-msft committed:
Add initial crate-version, rename package-dir and crate-name (#32687)

jhendrixMSFT committed:
make request and response media types configurable

moved query to block blob as it's specific to that blob type

jhendrixMSFT committed:
fix client naming for x-ms-meta and x-ms-or headers for Rust

jhendrixMSFT committed:
fix headers that contain time stamps

jhendrixMSFT committed:
consolidate redundant response headers

jhendrixMSFT committed:
fix getBlockList

cperaltah committed:
suppress swagger lint error

cperaltah committed:
update namespace structure + client customizations

cperaltah committed:
openapi

cperaltah committed:
fix service operations

cperaltah committed:
clean up routes

cperaltah committed:
update suppression

cperaltah committed:
update refs to blob namespace

jhendrixMSFT committed:
nest append/block/page blobs under container

jhendrixMSFT committed:
Use uint64 in Rust for content length headers

jhendrixMSFT committed:
fix client names for Rust

cperaltah committed:
regen openapi

cperaltah committed:
fix routes

cperaltah committed:
manually add slash to paths for ci

jhendrixMSFT committed:
fix bad application of clientName on alias

vincenttran-msft committed:
Correct package name from plural to singular

cperaltah committed:
revert openapi edit

jhendrixMSFT committed:
fix base64 encoded block IDs

jhendrixMSFT committed:
fix routes redux

jhendrixMSFT committed:
Rename path param blob to blobName

The previous name was a holdover from the swagger. The SDKs use blobName
as the parameter name.

cperaltah committed:
mark unwanted headers as internal for clients

cperaltah committed:
regen openapi

cperaltah committed:
suppress examples error

cperaltah committed:
manual swagger path update

jhendrixMSFT committed:
Move append/block/page blobs to children of blob

Revert previous change to DateResponseHeader

jhendrixMSFT committed:
don't use URL encoding for block ID

vincenttran-msft committed:
[Storage] Omit response headers for certain headers (#33593)

jhendrixMSFT committed:
Disable doc links for Rust codegen

Temporarily switched multipart/mixed to application/octec-stream until
the Rust emitter supports multipart/mixed.

catalinaperalta commented on specification/storage/Microsoft.BlobStorage/routes.tsp:
```diff
@@ -138,7 +138,7 @@ namespace Storage.Blob {
     #suppress "@typespec/http/content-type-ignored" "Template for existing API"
     @header("Content-Type")
-    contentType: "multipart/mixed";
+    contentType: "application/octet-stream";
```
@jhendrixMSFT didnt the original swagger specify multipart/mixed. Was that wrong?

jhendrixMSFT replied:
No. This is a temporary workaround until we have support for `multipart/mixed` in the Rust emitter (https://github.com/Azure/typespec-rust/issues/349).

jhendrixMSFT replied:
If there's a way to constrain this to just Rust, I can make that change.

jhendrixMSFT replied:
Maybe a better fix it to reinstate `multipart/mixed` and skip the offending operations for Rust. I'm looking into this now.

jhendrixMSFT replied:
I have this working, will push the change shortly.

jhendrixMSFT replied:
Done, see https://github.com/Azure/azure-rest-api-specs/commit/201a318efa3da6cc4dcf3e22c8e12ad3c43400eb

jhendrixMSFT committed:
Improvement on 4e80be1

vincenttran-msft committed:
Remove additionalDirectories from deps

vincenttran-msft committed:
Remove response headers from set_metadata (#33966)

vincenttran-msft committed:
[Storage] `set_properties` TypeSpec Change (#33939)

vincenttran-msft committed:
Omit response headers (#33970)

catalinaperalta commented on specification/storage/Microsoft.BlobStorage/routes.tsp:
```diff
@@ -665,8 +650,6 @@ namespace Storage.Blob {
         ...ImmutabilityPolicyModeResponseHeader;
         ...LegalHoldResponseHeader;
-        ...StructuredBodyTypeResponseHeader;
-        ...StructuredContentLengthResponseHeader;
```
@vincenttran-msft why were these headers removed? Does the service actually return these headers?

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
revert changes to package.json

cperaltah committed:
suppress record type issue

cperaltah committed:
modify openapi

cperaltah committed:
modify metadata header type

cperaltah committed:
openapi

cperaltah committed:
tsp format

cperaltah committed:
modify openapi

cperaltah committed:
suppress examples error

cperaltah committed:
suppress multipart issue

cperaltah committed:
fix suppression

cperaltah committed:
add suppression for base64 type

cperaltah committed:
clean up

cperaltah committed:
revert x-ms-meta x-ms-or header type

cperaltah committed:
fix breaking changes

cperaltah committed:
update openapi

cperaltah committed:
update metadata headers

cperaltah committed:
fix header

cperaltah committed:
openapi /

jhendrixMSFT committed:
update listContainersSegment to be pageable

v-jiaodi commented on specification/storage/Microsoft.BlobStorage/tspconfig.yaml:
```diff
@@ -0,0 +1,45 @@
+    flavor: azure
+  "@azure-tools/typespec-ts":
+    package-dir: "azure-storage-blob-rest"
+    generateMetadata: true
+    generateTest: true
+    packageDetails:
+      name: "@azure-rest/azure-storage-blob-rest"
```
```suggestion
    generate-metadata: true
    generate-test: true
    package-details:
      name: "@azure-rest/azure-storage-blob"
```

jhendrixMSFT committed:
Merge remote-tracking branch 'origin/main' into feature/blob-tsp

jhendrixMSFT committed:
update remaining pageable operations

vincenttran-msft committed:
Adjust route for start_copy_from_url (#34359)

cperaltah committed:
remove auth

cperaltah committed:
refactor

cperaltah committed:
add docs

jhendrixMSFT committed:
move xml.name decorator back to models

removed from response as it has no effect

jhendrixMSFT committed:
fix XML schema for list blobs operations

v-jiaodi committed:
Update specification/storage/Microsoft.BlobStorage/tspconfig.yaml

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
fix route issee and remove content type on no body

cperaltah committed:
wip

cperaltah committed:
fix no body ops

cperaltah committed:
Revert "wip"

This reverts commit 20bc5de1bb5d5e00dcac8bd2ae63bfa5e95034db.

cperaltah committed:
test swagger checks

jhendrixMSFT committed:
Add mediaTypeHint to StorageError response

Added missing Code field to StorageError.
Removed unused using statements.

vincenttran-msft committed:
[Storage] Fix extraneous `?<BlobType>` in RequestURI TypeSpec Changes (#35133)

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
Merge branch 'feature/blob-tsp' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

jhendrixMSFT committed:
make content-length a fixed value for some operations

When creating append and page blobs, the initial content-length must be
set to zero, no need for a parameter.

catalinaperalta committed:
Revert header removal (#35187)

* Revert "[Storage] Omit response headers for certain headers (#33593)"

This reverts commit 3a46f7cfede1e0a98a9ea42f4602e53e27761939.

* Revert "Remove response headers from set_metadata (#33966)"

This reverts commit dee1773d55d02440e68f8259a750dd45ec52e242.

* manual revert of 55c6629786cbecb492b481f6773a52e8bab3ac6e routes.tsp changes

* Revert "Omit response headers (#33970)"

This reverts commit a7ede2356175c7825fdb245ec17e5b6df71e1ae3.

* regen openapi

* fix server encrypted headers

* regen openapi

* remove extra clientName decorator

---------

Co-authored-by: catalinaperalta <caperal@microsoft.com>

catalinaperalta committed:
Update blob spec to use latest api version (#35231)

* update base api version

* add filerequestintent

* add new copy headers

* update storage error model

* update openapi spec

---------

Co-authored-by: catalinaperalta <caperal@microsoft.com>

cperaltah committed:
reset swagger to main

cperaltah committed:
update placeholder file

cperaltah committed:
fix more no body ops

cperaltah committed:
fix enums

cperaltah committed:
remove enum member ref

cperaltah committed:
add custom request id headers

cperaltah committed:
ci check

cperaltah committed:
add x-ms-error header

cperaltah committed:
fix immutability policy mode enum

cperaltah committed:
move scope decorator to client.tsp

cperaltah committed:
add lease action header

cperaltah committed:
regen openapi

cperaltah committed:
add missing headers

cperaltah committed:
add missing headers

cperaltah committed:
fix error headers

cperaltah committed:
ci

cperaltah committed:
fix more headers

cperaltah committed:
fix more headers and parms

cperaltah committed:
fix access tier

cperaltah committed:
fix header literal

cperaltah committed:
fix header enum

cperaltah committed:
fix response headers

cperaltah committed:
format + fix response header ref

cperaltah committed:
tsp format

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
regen openapi

vincenttran-msft committed:
[Storage] Move to `Access.internal` for Hiding Result Headers (#35254)

vincenttran-msft committed:
[Storage] `PageBlobClient` and `AppendBlobClient` initial introduction TypeSpec Changes (#35396)

vincenttran-msft committed:
[Storage] Lease operations and `AppendBlobClient` feature parity TypeSpec Changes (#35564)

* Make lease duration non-optional

* Adjust ResultHeaders for lease operations, rest of AppendBlobClient APIs

* Switch to privatization not removal

vincenttran-msft committed:
[Storage] Prep `.tsp` for `v0.3.0` release of Blob Storage SDK (#35886)

cperaltah committed:
regen openapi

cperaltah committed:
suppressions

cperaltah committed:
rename swagger

cperaltah committed:
reset swagger to main

cperaltah committed:
update readme

cperaltah committed:
update readme

cperaltah committed:
revert change

cperaltah committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/blob-tsp

cperaltah committed:
regen swagger

cperaltah committed:
suppressions

cperaltah committed:
suppressions

cperaltah committed:
add suppression

vincenttran-msft committed:
From ETag to Etag to match response body (#36377)

mikeharder committed:
Merge branch 'main' into feature/blob-tsp

mikeharder committed:
empty

cperaltah committed:
remove extra suppressions

vincenttran-msft committed:
[Storage] `get_page_ranges`, `update_sequence_number` and `upload_pages_from_url` TypeSpec Changes (#36311)

catalinaperalta commented on specification/storage/Microsoft.BlobStorage/routes.tsp:
```diff
@@ -1715,7 +1715,6 @@ namespace Storage.Blob {
             ...ContentMd5ResponseHeader;
             ...ContentCrc64ResponseHeader;
             ...BlobSequenceNumberResponseHeader;
-            ...DateResponseHeader;
```
@vincenttran-msft why did you remove the date headers in this comment? Shouldnt you be using the private version of the headers?

vincenttran-msft replied:
Good catch, I thought I had fixed this during the in-meeting review but must have been missed.

catalinaperalta commented on specification/storage/Microsoft.BlobStorage/models.tsp:
```diff
@@ -934,10 +934,14 @@ model BlobFlatListSegment {
-  @Xml.name("ClearRange") clearRange?: ClearRange[];
+  @Xml.unwrapped
+  @Xml.name("ClearRange")
+  clearRange?: ClearRange[];
```
Just confirming, are these properties really unwrapped in the xml response? Have you tested with an actual service response?

catalinaperalta commented:
@vincenttran-msft please regenerate the openapi spec based off of your recent commits. The typespec validation step is failing again.

vincenttran-msft committed:
[Storage] Regen after latest (#36594)

catalinaperalta commented:
@vincenttran-msft please look at the errors in the typespec validation check and address. You need to regenerate the openapi spec like I mentioned in my earlier comments. 

@jalauzon-msft let's continue discussing in today's sync with Scott :)

vincenttran-msft committed:
[Storage] Recompile after `.tsp` changes (#36649)

vincenttran-msft committed:
Merge branch 'main' into feature/blob-tsp
